File: skeleton/src/main/java/me/shkschneider/skeleton/helper/ApplicationHelper.java
package me.shkschneider.skeleton.helper;

import android.annotation.TargetApi;
import android.app.Activity;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.Signature;
import android.content.res.AssetManager;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.provider.Settings;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import me.shkschneider.skeleton.BuildConfig;
import me.shkschneider.skeleton.ImageManipulator;
import me.shkschneider.skeleton.SkeletonApplication;

public class ApplicationHelper {

    public static boolean debug() {
        // Workaround to get DEBUG value of the android application and not that android-library
        try {
            final Class<?> cls = Class.forName(packageName() + ".BuildConfig");
            final Field field = cls.getField("DEBUG");
            return (Boolean) field.get(false);
        }
        catch (final Exception e) {
            LogHelper.wtf(e);
        }
        return BuildConfig.DEBUG;
    }

    public static Resources resources() {
        return SkeletonApplication.CONTEXT.getResources();
    }

    public static AssetManager assets() {
        return SkeletonApplication.CONTEXT.getAssets();
    }

    public static String packageName() {
        return SkeletonApplication.CONTEXT.getPackageName();
    }

    public static PackageManager packageManager() {
        return SkeletonApplication.CONTEXT.getPackageManager();
    }

    public static String name() {
        try {
            final PackageManager packageManager = packageManager();
            if (packageManager == null) {
                LogHelper.warning("PackageManager was NULL");
                return null;
            }

            final ApplicationInfo applicationInfo = packageManager.getApplicationInfo(packageName(), 0);
            if (applicationInfo == null) {
                LogHelper.warning("ApplicationInfo was NULL");
                return null;
            }

            final CharSequence label = applicationInfo.loadLabel(packageManager);
            if (label == null) {
                LogHelper.warning("Label was NULL");
                return null;
            }

            return label.toString();
        }
        catch (final Exception e) {
            LogHelper.wtf(e);
            return null;
        }
    }

    public static String versionName() {
        try {
            final PackageManager packageManager = packageManager();
            if (packageManager == null) {
                LogHelper.warning("PackageManager was NULL");
                return null;
            }

            return packageManager.getPackageInfo(packageName(), PackageManager.GET_META_DATA).versionName;
        }
        catch (final Exception e) {
            LogHelper.wtf(e);
            return null;
        }
    }

    public static int versionCode() {
        try {
            final PackageManager packageManager = packageManager();
            if (packageManager == null) {
                LogHelper.warning("PackageManager was NULL");
                return -1;
            }

            return packageManager.getPackageInfo(packageName(), PackageManager.GET_META_DATA).versionCode;
        }
        catch (final Exception e) {
            LogHelper.wtf(e);
            return -1;
        }
    }

    public static Bitmap icon() {
        try {
            final PackageManager packageManager = packageManager();
            if (packageManager == null) {
                LogHelper.warning("PackageManager was NULL");
                return null;
            }

            final ApplicationInfo applicationInfo = packageManager.getApplicationInfo(packageName(), 0);
            if (applicationInfo == null) {
                LogHelper.warning("ApplicationInfo was NULL");
                return null;
            }

            final Drawable drawable = applicationInfo.loadIcon(packageManager);
            return ImageManipulator.fromDrawable(drawable);
        }
        catch (final Exception e) {
            LogHelper.wtf(e);
            return null;
        }
    }

    public static List<String> permissions() {
        try {
            final PackageManager packageManager = packageManager();
            if (packageManager == null) {
                LogHelper.warning("PackageManager was NULL");
                return null;
            }

            final PackageInfo packageInfo = packageManager.getPackageInfo(packageName(), PackageManager.GET_PERMISSIONS);

            final List<String> list = new ArrayList<String>();
            Collections.addAll(list, packageInfo.requestedPermissions);
            return list;
        }
        catch (final Exception e) {
            LogHelper.wtf(e);
            return null;
        }
    }

    public static String signature() {
        final PackageManager packageManager = packageManager();
        if (packageManager == null) {
            LogHelper.warning("PackageManager was NULL");
            return null;
        }

        try {
            final PackageInfo packageInfo = packageManager.getPackageInfo(packageName(), PackageManager.GET_SIGNATURES);
            if (packageInfo == null) {
                LogHelper.warning("PackageInfo was NULL");
                return null;
            }

            final Signature[] signatures = packageInfo.signatures;
            if (signatures == null) {
                LogHelper.debug("No signatures");
                return null;
            }

            return signatures[0].toCharsString();
        }
        catch (final Exception e) {
            LogHelper.wtf(e);
            return null;
        }
    }

    public static boolean fromMarket() {
        if (AndroidHelper.api() >= AndroidHelper.API_21) {
            return fromMarket21();
        }
        else if (AndroidHelper.api() >= AndroidHelper.API_17) {
            return fromMarket17();
        }
        return fromMarket3();
    }

    @TargetApi(AndroidHelper.API_21)
    public static boolean fromMarket21() {
        // API-21+ Settings.Secure
        return (Settings.Secure.getInt(SkeletonApplication.CONTEXT.getContentResolver(), Settings.Secure.INSTALL_NON_MARKET_APPS, 0) == 1);
    }

    @TargetApi(AndroidHelper.API_17)
    @SuppressWarnings("deprecation")
    public static boolean fromMarket17() {
        // API-17+ Settings.Global
        return (Settings.Global.getInt(SkeletonApplication.CONTEXT.getContentResolver(), Settings.Global.INSTALL_NON_MARKET_APPS, 0) == 1);
    }

    @TargetApi(AndroidHelper.API_3)
    public static boolean fromMarket3() {
        // API-3+ Settings.Secure
        return (Settings.Secure.getInt(SkeletonApplication.CONTEXT.getContentResolver(), Settings.Secure.INSTALL_NON_MARKET_APPS, 0) == 1);
    }

}

File: core/src/main/kotlin/me/shkschneider/skeleton/helper/ApplicationHelper.kt
package me.shkschneider.skeleton.helper

import android.annotation.SuppressLint
import android.annotation.TargetApi
import android.content.pm.ApplicationInfo
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.content.res.Resources
import android.graphics.drawable.Drawable
import android.os.Build
import me.shkschneider.skeleton.SkeletonApplication
import me.shkschneider.skeleton.helperx.Logger
import me.shkschneider.skeleton.helperx.SystemServices

object ApplicationHelper {

    // ApplicationInfo.loadDefaultIcon(packageManager)
    const val DEFAULT_ICON = android.R.drawable.sym_def_app_icon
    const val INSTALLER = "com.android.vending"

    fun debuggable(): Boolean {
        // <https://stackoverflow.com/a/25517680/603270>
        return SkeletonApplication.DEBUGGABLE
    }

    fun resources(): Resources {
        // Resources.getSystem()
        return ContextHelper.applicationContext().resources
    }

    fun packageName(): String {
        return ContextHelper.applicationContext().packageName
    }

    fun packageManager(): PackageManager {
        return ContextHelper.applicationContext().packageManager
    }

    fun applicationInfo(packageName: String = ApplicationHelper.packageName()): ApplicationInfo? {
        try {
            return packageManager().getApplicationInfo(packageName, 0)
        } catch (e: PackageManager.NameNotFoundException) {
            Logger.wtf(e)
            return null
        }
    }

    fun name(packageName: String = ApplicationHelper.packageName()): String? {
        applicationInfo(packageName)?.loadLabel(packageManager())?.let { label ->
            return label.toString()
        }
        Logger.warning("Label was NULL")
        return null
    }

    fun exists(packageName: String = ApplicationHelper.packageName()): Boolean {
        try {
            return (packageManager().getApplicationInfo(packageName, 0) != null)
        } catch (e: PackageManager.NameNotFoundException) {
            Logger.wtf(e)
            return false
        }
    }

    fun packageInfo(packageName: String = ApplicationHelper.packageName(), flags: Int = 0): PackageInfo? {
        try {
            val packageManager = packageManager()
            return packageManager.getPackageInfo(packageName, flags)
        } catch (e: PackageManager.NameNotFoundException) {
            Logger.wtf(e)
            return null
        }
    }

    fun versionName(packageName: String = ApplicationHelper.packageName()): String? {
        return packageInfo(packageName, PackageManager.GET_META_DATA)?.versionName ?: run {
            Logger.warning("VersionName was NULL")
            return null
        }
    }

    @SuppressLint("NewApi")
    fun versionCode(packageName: String = ApplicationHelper.packageName()): Long? {
        val packageInfo = packageInfo(packageName, PackageManager.GET_META_DATA)
        if (Build.VERSION.SDK_INT >= 28) {
            return packageInfo?.longVersionCode ?: run {
                Logger.warning("LongVersionCode was NULL")
                return null
            }
        } else {
            @Suppress("DEPRECATION")
            return packageInfo?.versionCode?.toLong() ?: run {
                Logger.warning("VersionCode was NULL")
                return null
            }
        }
    }

    fun buildTime(packageName: String = ApplicationHelper.packageName()): Long? {
        return packageInfo(packageName, PackageManager.GET_META_DATA)?.lastUpdateTime ?: run {
            Logger.warning("LastUpdateTime was NULL")
            return null
        }
    }

    fun permissions(packageName: String = ApplicationHelper.packageName()): List<String>? {
        return packageInfo(packageName, PackageManager.GET_PERMISSIONS)?.requestedPermissions?.toList()
                ?: run {
                    Logger.warning("RequestedPermissions was NULL")
                    return null
                }
    }

    fun icon(packageName: String = ApplicationHelper.packageName()): Int? {
        try {
            val packageManager = packageManager()
            val applicationInfo = packageManager.getApplicationInfo(packageName, 0)
            return applicationInfo.icon
        } catch (e: PackageManager.NameNotFoundException) {
            Logger.wtf(e)
            return null
        }
    }

    fun drawable(packageName: String = ApplicationHelper.packageName()) : Drawable? {
        try {
            val packageManager = packageManager()
            val applicationInfo = packageManager.getApplicationInfo(packageName, 0)
            return applicationInfo.loadIcon(packageManager)
        } catch (e: PackageManager.NameNotFoundException) {
            Logger.wtf(e)
            return null
        }
    }

    @TargetApi(AndroidHelper.API_19)
    fun clear() {
        Logger.info("clearApplicationUserData()")
        SystemServices.activityManager()?.clearApplicationUserData()
    }

}
